apps/prairielearn/src/ee/lib/billing/components/InstructorInstanceAdminBillingForm.html.ts

```diff
@@ -0,0 +1,257 @@
+import { PlanName, planGrantsMatchPlanFeatures } from '../plans-types';
+
+interface InstructorInstanceAdminBillingInput {
+  initialRequiredPlans: PlanName[];
+  desiredRequiredPlans: PlanName[];
+  institutionPlanGrants: PlanName[];
+  courseInstancePlanGrants: PlanName[];
+  enrollmentCount: number;
+  enrollmentLimit: number;
+}
+
+interface InstructorInstanceAdminBillingState {
+  studentBillingEnabled: boolean;
+  studentBillingCanChange: boolean;
+  studentBillingDidChange: boolean;
+  studentBillingAlert: AlertProps | null;
+  computeEnabled: boolean;
+  computeCanChange: boolean;
+  computeDidChange: boolean;
+  computeAlert: AlertProps | null;
+}
+
+interface AlertProps {
+  message: string;
+  color: 'success' | 'warning' | 'danger';
+}
+
+export function instructorInstanceAdminBillingState(
+  input: InstructorInstanceAdminBillingInput
+): InstructorInstanceAdminBillingState {
+  const studentBillingInitialEnabled = input.initialRequiredPlans.includes('basic');
+  const studentBillingEnabled = input.desiredRequiredPlans.includes('basic');
+  const computeEnabledByInstitution = planGrantsMatchPlanFeatures(
+    input.institutionPlanGrants,
+    'compute'
+  );
+  const computeEnabledByCourseInstance = planGrantsMatchPlanFeatures(
+    input.courseInstancePlanGrants,
+    'compute'
+  );
+  const computeEnabled =
+    (!studentBillingEnabled && (computeEnabledByInstitution || computeEnabledByCourseInstance)) ||
+    input.desiredRequiredPlans.includes('compute');
+
+  let studentBillingCanChange = true;
+  const studentBillingDidChange =
+    input.initialRequiredPlans.includes('basic') !== input.desiredRequiredPlans.includes('basic');
+  let studentBillingAlert: AlertProps | null = null;
+  if (studentBillingInitialEnabled && input.enrollmentCount > input.enrollmentLimit) {
+    studentBillingCanChange = false;
+    const inflectedCountVerb = input.enrollmentCount === 1 ? 'is' : 'are';
+    const inflectedCountNoun = input.enrollmentCount === 1 ? 'enrollment' : 'enrollments';
+    studentBillingAlert = {
+      message: [
+        `There ${inflectedCountVerb} ${input.enrollmentCount} ${inflectedCountNoun} in this course, which exceeds the limit of ${input.enrollmentLimit}.`,
+        'To disable student billing, first remove excess enrollments.',
+      ].join(' '),
+      color: 'warning',
+    };
+  }
+
+  let computeCanChange = true;
+  const computeDidChange =
+    input.initialRequiredPlans.includes('compute') !==
+    input.desiredRequiredPlans.includes('compute');
+  let computeAlert: AlertProps | null = null;
+  if (!studentBillingEnabled && (computeEnabledByInstitution || computeEnabledByCourseInstance)) {
+    computeCanChange = false;
+    computeAlert = {
+      message: computeEnabledByCourseInstance
+        ? 'This course instance already has access to compute features without additional payment.'
+        : 'Courses in your institution already have access to compute features without additional payment.',
```

Comment:
Will folks care about this distinction in practice?

Reply:
This is handling the case where we add a plan grant to the course instance itself, not to the parent institution. Recall that a course instance _plan grant_ is different from a course instance _required plan_ - the former makes features available, the latter requires students to have access to the feature (possibly through student-pays) to access the course instance.

If this is confusing to you, it's almost definitely confusing to end users. Maybe the only message we need is "This course instance already has access to compute features without additional payment"?

Reply:
Right, thanks for the reminder! And yes, I think we should focus on the important information for the instructor as you say.

Reply:
I simplified this to a single message in 828cfa7.